traitlets/traitlets.py

```diff
@@ -1342,6 +1455,14 @@ def add_traits(self, **traits):
         """Dynamically add trait attributes to the HasTraits instance."""
         self.__class__ = type(self.__class__.__name__, (self.__class__,),
                               traits)
+
+        events = self._static_trait_notifiers
+        events.extend(self.trait_events().values())
+
+        for e in events:
+            if e.caches_instance_resources:
```

Comment:
Similarly to `_static_trait_notifiers` I need to rethink this and make the implementation cleaner.